### ISSUE TITLE
[1.12rc4] Occasionally Swarm cluster does not work with 2 of 3 manager nodes online

### ISSUE THREAD

**Issue by Stehsegler** [docker/docker#25055]
**Output of `docker version`:**

```
Client:
 Version:      1.12.0-rc4
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   e4a0dbc
 Built:
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0-rc4
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   e4a0dbc
 Built:
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 6
 Running: 0
 Paused: 0
 Stopped: 6
Images: 7
Server Version: 1.12.0-rc4
Storage Driver: devicemapper
 Pool Name: vg--docker-data
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file:
 Metadata file:
 Data Space Used: 5.386 GB
 Data Space Total: 52.61 GB
 Data Space Available: 47.22 GB
 Metadata Space Used: 5.366 MB
 Metadata Space Total: 532.7 MB
 Metadata Space Available: 527.3 MB
 Thin Pool Minimum Free Space: 5.261 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Library Version: 1.02.107-RHEL7 (2016-06-09)
Logging Driver: journald
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host overlay null
Swarm: pending
 NodeID: xxx
 IsManager: Yes
 Managers: 0
 Nodes: 0
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.22.2.el7.x86_64
Operating System: Red Hat Enterprise Linux Server 7.2 (Maipo)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.64 GiB
Name: xxx
ID: xxx
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: xxx
No Proxy: xxx
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Red Hat 7 on VMware ESX

**Steps to reproduce the issue:**
1. Create a Swarm cluster with three manager nodes
2. Shutdown all three Docker Daemon instances
3. Start two Docker Daemon instances

**Describe the results you received:**
Occasionally, the cluster doesn't get into a functional state with two nodes, i.e. services are not started and trying to use cluster functionality (e.g. `docker node ls`) leads to `Error response from daemon: rpc error: code = 4 desc = context deadline exceeded`.

The daemons log something like this in an endless loop:

Node 1:
```
Jul 26 11:58:08 xxx docker[20662]: time="2016-07-26T11:58:08.575594001+02:00" level=info msg="3af9378ad3ec748d is starting a new election at term 16733"
Jul 26 11:58:08 xxx docker[20662]: time="2016-07-26T11:58:08.575679688+02:00" level=info msg="3af9378ad3ec748d became candidate at term 16734"
Jul 26 11:58:08 xxx docker[20662]: time="2016-07-26T11:58:08.575698825+02:00" level=info msg="3af9378ad3ec748d received vote from 3af9378ad3ec748d at term 16734"
Jul 26 11:58:08 xxx docker[20662]: time="2016-07-26T11:58:08.575732653+02:00" level=info msg="3af9378ad3ec748d [logterm: 16702, index: 7046] sent vote request to 40ce655e1dd06f7 at term 16734"
Jul 26 11:58:08 xxx docker[20662]: time="2016-07-26T11:58:08.575753411+02:00" level=info msg="3af9378ad3ec748d [logterm: 16702, index: 7046] sent vote request to 599ad5c0cefa03b8 at term 16734"
Jul 26 11:58:11 xxx docker[20662]: time="2016-07-26T11:58:11.610368028+02:00" level=error msg="agent: session failed" error="session initiation timed out" module=agent
```

Node 2:
```
Jul 26 11:55:34 xxx docker[10023]: time="2016-07-26T11:55:34.130782155+02:00" level=info msg="40ce655e1dd06f7 [term: 16728] received a MsgVote message with higher term from 3af9378ad3ec748d [term: 16729]"
Jul 26 11:55:34 xxx docker[10023]: time="2016-07-26T11:55:34.130842529+02:00" level=info msg="40ce655e1dd06f7 became follower at term 16729"
Jul 26 11:55:34 xxx docker[10023]: time="2016-07-26T11:55:34.130866766+02:00" level=info msg="40ce655e1dd06f7 [logterm: 16702, index: 7046, vote: 0] voted for 3af9378ad3ec748d [logterm: 16702, index: 7046] at term 16729"
Jul 26 11:55:35 xxx docker[10023]: time="2016-07-26T11:55:35.572655684+02:00" level=error msg="agent: session failed" error="session initiation timed out" module=agent
```

**Describe the results you expected:**
The cluster gets in a functional state with two nodes, i.e. services are started and cluster functionality like `docker service ls` can be used. Two of three nodes should be enough to reach consensus.

**Additional information you deem important (e.g. issue happens only occasionally):**
Happens to me around every second to third cluster startup.

When the third node is started, the cluster becomes functional. However, when the cluster started up like this (consensus not reached before all three nodes were available) and I shutdown the Leader node afterwards, the cluster loses its functionality again, until the third node is once more started. When stopping a non-Leader node, however, the cluster stays functional.

**Comment by thaJeztah** [comment]
ping @LK4D4 @aaronlehmann

**Comment by Stehsegler** [comment]
Experiencing similar behaviour also with 1.12.0-rc5, however with a slightly different error message when trying to use cluster functionality: `Error response from daemon: rpc error: code = 2 desc = raft: no elected cluster leader`

**Comment by abronan** [comment]
Spent some time investigating on this and could reproduce.

Considering *Node 1*, *Node 2* and *Node 3*, shutting down the whole cluster, and restarting only *Node 1* and *Node 2*.

**Node 1** :

```
INFO[0070] f54e5424c276d82 is starting a new election at term 13
INFO[0070] f54e5424c276d82 became candidate at term 14
INFO[0070] f54e5424c276d82 received vote from f54e5424c276d82 at term 14
INFO[0070] f54e5424c276d82 [logterm: 8, index: 135] sent vote request to 371429e320cb4749 at term 14
INFO[0070] f54e5424c276d82 [logterm: 8, index: 135] sent vote request to 368efe730d913751 at term 14

// We should receive the vote from Node 2 here, but we don't thus waiting for the election timeout (5s here) to trigger a new election

INFO[0076] f54e5424c276d82 is starting a new election at term 14
INFO[0076] f54e5424c276d82 became candidate at term 15
INFO[0076] f54e5424c276d82 received vote from f54e5424c276d82 at term 15
INFO[0076] f54e5424c276d82 [logterm: 8, index: 135] sent vote request to 371429e320cb4749 at term 15
INFO[0076] f54e5424c276d82 [logterm: 8, index: 135] sent vote request to 368efe730d913751 at term 15

// Goes on and on...
```

**Node 2** :

```
INFO[0069] 371429e320cb4749 [term: 13] received a MsgVote message with higher term from f54e5424c276d82 [term: 14]
INFO[0069] 371429e320cb4749 became follower at term 14
INFO[0069] 371429e320cb4749 [logterm: 8, index: 135, vote: 0] voted for f54e5424c276d82 [logterm: 8, index: 135] at term 14

// At this point, we voted for Node 1 but it never receives the vote

ERRO[0070] agent: session failed error=session initiation timed out module=agent
INFO[0075] 371429e320cb4749 [term: 14] received a MsgVote message with higher term from f54e5424c276d82 [term: 15]
INFO[0075] 371429e320cb4749 became follower at term 15
INFO[0075] 371429e320cb4749 [logterm: 8, index: 135, vote: 0] voted for f54e5424c276d82 [logterm: 8, index: 135] at term 15

// We keep voting for Node 1 and waiting for an acknowledgment of its Leader status.
```

The apparent cause is that **Node 1** *does not* receive votes from **Node 2**.

My first findings is that **Node 2** occasionally does not trigger the `Ready` state when it has to send messages from its mailbox to other nodes. Thus it votes internally for its state machine but does not send the vote for `Node 1`. My first guess was that something could be wrong with the connection but both nodes would communicate just fine before it would notify that it dropped the leader.

That's also why restarting **Node 3** could unblock the situation: because it gets a majority and elects a Leader. Restarting **Node 1** and **Node 2** could also fix the issue. Seems like a race on initialization or the election.

Another observation is that bumping the election timeout to `5s` or `10s`, the issue is hardly reproducible.

----

In comparison, here is a successful election on restart for **Node 1** and **Node 2** with **Node 3** being down:

**Node 1**
```
INFO[0010] 582c953040438194 is starting a new election at term 33 
INFO[0010] 582c953040438194 became candidate at term 34 
INFO[0010] 582c953040438194 received vote from 582c953040438194 at term 34 
INFO[0010] 582c953040438194 [logterm: 9, index: 194] sent vote request to 100f5a3d0c014a2d at term 34 
INFO[0010] 582c953040438194 [logterm: 9, index: 194] sent vote request to 1b3b9b42ae0d5dd4 at term 34 
INFO[0010] 582c953040438194 received vote from 100f5a3d0c014a2d at term 34 
INFO[0010] 582c953040438194 [quorum:2] has received 2 votes and 0 vote rejections 
INFO[0010] 582c953040438194 became leader at term 34
```

**Node 2**
```
INFO[0009] 100f5a3d0c014a2d [term: 10] received a MsgVote message with higher term from 582c953040438194 [term: 34] 
INFO[0009] 100f5a3d0c014a2d became follower at term 34  
INFO[0009] 100f5a3d0c014a2d [logterm: 9, index: 194, vote: 0] voted for 582c953040438194 [logterm: 9, index: 194] at term 34 
INFO[0009] raft.node: 100f5a3d0c014a2d elected leader 582c953040438194 at term 34
```
---

Still working on finding the root cause, but as of now restarting the first two nodes should recover the cluster.

**Comment by abronan** [comment]
Update: vendoring etcd's raft v3 seems to solve the issue. Can't reproduce including the new version.

**Comment by thaJeztah** [comment]
@abronan is that vendored in 1.12.0, or you only tried in a local build?

if not in 1.12.0, should we fix this in a patch release?

**Comment by aaronlehmann** [comment]
I discussed this with a few people yesterday and I think we should vendor the latest etcd/raft library in 1.12.1. Ideally we would only take a specific commit that solves the problem, but I don't think we know which one solves it. Bisecting would be difficult because the issue can't be triggered reliably.

**Comment by thaJeztah** [comment]
Let me add this to the 1.12.1 milestone for consideration

**Comment by Stehsegler** [comment]
IMHO, it is really important to have this issue fixed quickly. The background of the multi-manager support in Swarm is mainly fault tolerance, so reliability is really a major point here. As described in the initial post, the issue is not only about not being able to startup the cluster with only 2 of 3 managers, but also about the it losing functionality when the leader is shut down after such an "unclean" startup procedure (i.e. consensus not reached before all three nodes are up). Thus it is possible (and not quite unlikely) that a cluster with three managers which is supposed to tolerate shutdown/crash of an arbitrary node does not do so, without any real possibility for the user to notice that the cluster is in such kind of dangerous state, unless he knows about the issue and re-checks every time the cluster as a whole is restarted.

**Comment by aaronlehmann** [comment]
Possible fix at https://github.com/docker/swarmkit/pull/1310

**Comment by aaronlehmann** [comment]
Closing, since we vendored a fix that we believe solves the problem. Please reopen if the issue reoccurs with docker master, or the upcoming 1.12.1 release.
